File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int hashMap[26] = {0};
        int result = 0, max_freq = 0;

        for(int start = 0, end = 0; end < s.size(); end++){
            hashMap[s[end] - 'A']++;
            max_freq = max(max_freq, hashMap[s[end] - 'A']);

            while((end - start + 1) - max_freq > k){
                hashMap[s[start] - 'A']--;
                start++;
            }

            result = max(result, end - start + 1);
        }

        return result;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        hashMap = defaultdict(int)
        start = 0
        max_len = 0
        max_freq = 0

        for end in range(len(s)):
            hashMap[s[end]] += 1
            max_freq = max(max_freq, hashMap[s[end]])

            while (end - start + 1) - max_freq > k:
                hashMap[s[start]] -= 1
                start += 1

            max_len = max(max_len, end - start + 1)

        return max_len